src/drivers/esp32/src/esp_slip.c

```diff
@@ -78,6 +98,144 @@ static void sendSlipPacket(uint32_t size, uint8_t *data, coms_sendbuffer_t sendB
     sendBufferFn(dmaSendSize, &dmaSendBuffer[0]);
   }
 }
+
+static slipDecoderStatus_t decodeSlipPacket(uint8_t c, esp_slip_receive_packet *receiverPacket, esp_slip_send_packet *senderPacket)
+{
+
+  slipDecoderStatus_t decoderStatus = SLIP_DECODING;
+  switch (espblReceiveState)
+  {
+  case receiveStart:
+    receiverPacket->status = 1;
+    espblReceiveState = (c == 0xC0) ? receiveDirection : receiveStart;
+    break;
+  case receiveDirection:
+    if (c == 0x01)
+    {
+      espblReceiveState = receiveCommand;
+    }
+    else
+    {
+      espblReceiveState = error;
+      decoderStatus = SLIP_ERROR;
+    }
+    break;
+  case receiveCommand:
+    receiverPacket->command = c;
+    slipSizeIndex = 0;
+    if (c == senderPacket->command)
+    {
+      espblReceiveState = receiveSize;
+    }
+    else
+    {
+      espblReceiveState = error;
+      decoderStatus = SLIP_ERROR;
+    }
+    break;
+  case receiveSize:
+    slipSize[slipSizeIndex] = c;
+    if (slipSizeIndex == 1)
+    {
+      receiverPacket->dataSize = ((uint16_t)slipSize[0] + ((uint16_t)slipSize[1] << 8));
+      slipValueIndex = 0;
+      if (receiverPacket->dataSize > 0 && receiverPacket->dataSize < ESP_MTU)
+      {
+        espblReceiveState = receiveValue;
+      }
+      else
+      {
+        espblReceiveState = error;
+        decoderStatus = SLIP_ERROR;
+      }
+    }
+    slipSizeIndex++;
+    break;
+  case receiveValue: // only used for READ_REG
+    slipValue[slipValueIndex] = c;
+    if (slipValueIndex == 3)
+    {
+      receiverPacket->value = ((uint32_t)slipValue[0] + ((uint32_t)slipValue[1] << 8) + ((uint32_t)slipValue[2] << 16) + ((uint32_t)slipValue[3] << 24));
```

Comment:
This is begging for a comment, it might be obvious to Rik today, but not to Rik tomorrow, or Jonas today :)

```
//
// Explaination of this subscripted bitshifted genious code
//
receiverPacket->value =  ((uint32_t) slipValue[0] + ((uint32_t) slipValue[1] << 8);
receiverPacket->value += ((uint32_t) slipValue[2] << 16) + ((uint32_t) slipValue[3] << 24));
```